flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -0,0 +1,124 @@
+//===-- Lower/Support/BoxValue.h -- FIR box values --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LOWER_SUPPORT_BOXVALUE_H
+#define LOWER_SUPPORT_BOXVALUE_H
+
+#include "mlir/IR/Value.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace fir {
+
+//===----------------------------------------------------------------------===//
+//
+// Boxed values
```

Comment:
I used the term box here because it is serving an identical purpose in a different context. The bridge must track a variable and its attributes (all possibly runtime values). To date, the bridge does it implicitly in the IR itself as either the attributes are constants or tracked by emboxing and unboxing pairs of values. That approach is going to scale poorly.
An expression tree can contain array-typed values. Whether those attributes are reified in the IR as a box or maintained in the bridge as an abstraction, the bridge will necessarily track those dictionaries.
The dynamic cases you refer to are exactly the motivation. As the array is mutated by the surrounding context, it's shape (tracking dynamically computed values) can be maintained and passed back to the previous node of the tree.
(I'll try to fix my description, since I agree it is misleading.)